vision.md

```diff
@@ -1,115 +1,407 @@
+    - Unrecoverable errors may still cause the binary to shutdown eventually,
+      but not unexpectedly.
+1. The core implementation is written in Rust
+    - Plugins that connect to a thin-edge.io binary, may be written in another
+      programming language
+1. A deployment of thin-edge.io consists of a single binary with a single
+   configuration entry-point
+    - The final configuration may consists of more than one resource (file,
+      network resource, etc), and potentially be even loaded over the network,
+      this is left open
+1. The configuration is the single point of truth w.r.t. the initial state of
+   the components mentioned within
+    1. The default state of a plugin must be documented
+    1. A component may use a (documented) default value for a missing
+       configuration entry
+1. The configuration is the single point of truth for the communication between
+   components inside a single thin-edge.io binary deployment
+1. The communication between the components is verified to be compatible in
+   advance
+1. The out- and in-bound connectivity is mediated through a framework specific
+   format
+    - JSON is the lingua franca, but other forms may be acceptable if they stay
+      within the capabilites of JSON
+    - json-schema is used to document the (JSON) format in a machine-readable
+      way
+1. The principal way of extending thin-edge.io is over well-defined Rust interfaces (traits and other types)
+    - This does not preclude other forms of extensions in other languages (i.e.
+      a bridge over MQTT to python)
+1. Starting a thin-edge.io binary for development use or production use is not
+   interactive
+    - Starting a thin-edge.io binary for setup or similar purposes may be
+      interactive
+1. Components can persist data using the framework
+    - to persist data between restarts of the deployment
+    - to cache data during network outages
+    - to provide operation checkpoints during sensitive operations
+1. User-configurable logging is provided via the framework
+    - A user is able to configure logging per-component as well as globally for
+      the framework
+
+### Capabilities
+
+<!--
+The following list of capabilities should be considered non-final. It might be
+extended in the future.
+-->
+
+In combination with IoT platforms, thin-edge.io is a foundation for enabling
+devices with the following capabilities:
+
+1. Connectivity to the major cloud providers in the IIOT space
+    - for effortless and secure edge device lifecycle management for single
+      and device fleets
+    - for low-touch provisioning of thin edge devices
+    - for local and remote configuration
+    - for local and remote maintenance including remote access
+      (monitoring/troubleshooting)
+    - decommissioning of thin devices (e.g. for security compromised or end of
+      life devices)
+1. "South bound" connectivity to devices via the major protocols in the IIOT industry
+1. On-device data preprocessing
```

Comment:
Similar to above: If we call those feautres out we should make clear that we are just making it easy to leverage exisiitng analytics solutions but not develop them ourselfes.